### PR TITLE
[rust] [rust-server] Add AbstractRustCodegen

### PR DESCRIPTION
**Pull request by jacob-pro**
Please review when you have chance: @wing328 @spacether @richardwhiuk @paladinzh 

This creates an AbstractRustCodegen (and unit tests)

The primary benefit is that identifier handling has been made more reliable, fixing various inconsistencies, as well as removing about 200 lines code duplication.
This will also be useful in future if we want to add support for additional Rust code generators / frameworks.

I have put in a lot of time and effort to make the effects of this change as small as possible.

However there are a few subtle breaking changes to identifier naming: 
- Reserved words are escaped differently (closes https://github.com/OpenAPITools/openapi-generator/issues/13171)
- **Enum variants are now converted to camel case** instead of uppercase (as per guidelines https://rust-lang.github.io/api-guidelines/naming.html)
- More special characters are now mapped e.g. `$$` becomes `DollarDollar` (see https://github.com/OpenAPITools/openapi-generator/pull/10521#issuecomment-941949318)

If any of those changes are going to be an issue for either of the generators let me know and I will add in specific overrides to cancel them out?

Thanks in advance for your time reviewing :)

(See also [slack thread](https://openapi-generator.slack.com/archives/CLSB0U0R5/p1660559118652119?thread_ts=1660409047.365439&cid=CLSB0U0R5) about creating this abstract class)

Example log output:

```
[main] WARN org.openapitools.codegen.languages.AbstractRustCodegen - 1Pending cannot be used as a enum variant name. Renamed to Variant1Pending
[main] WARN org.openapitools.codegen.languages.AbstractRustCodegen - 1_pet_summary cannot be used as a model file name. Renamed to model_1_pet_summary
[main] WARN org.openapitools.codegen.languages.AbstractRustCodegen - 1PetSummary cannot be used as a model name. Renamed to Model1PetSummary
[main] WARN org.openapitools.codegen.languages.AbstractRustCodegen - 1_create_pet cannot be used as a method name. Renamed to call_1_create_pet
[main] WARN org.openapitools.codegen.languages.AbstractRustCodegen - 1_pet_name cannot be used as a parameter name. Renamed to param_1_pet_name
[main] WARN org.openapitools.codegen.languages.AbstractRustCodegen - 1_pet_name cannot be used as a field/variable name. Renamed to param_1_pet_name
```

### CODE REVIEW

**Comment by wing328** [comment]
> Enum variants are now converted to camel case instead of uppercase

sounds like that's a breaking change (not backward-compatible) right?

**Comment by wing328** [comment]
FYI. To run the tests locally, you will need to run a local petstore server: https://github.com/OpenAPITools/openapi-generator/wiki/Integration-Tests#how-to-add-integration-tests-for-new-petstore-samples

**Comment by wing328** [comment]
cc @frol (2017/07) @farcaller (2017/08) @richardwhiuk (2019/07) @paladinzh (2020/05)

File: samples/server/petstore/rust-server/output/openapi-v3/src/models.rs [hunk 1]
```diff
@@ -629,16 +629,16 @@ impl AnyOfObject {
 #[cfg_attr(feature = "conversion", derive(frunk_enum_derive::LabelledGenericEnum))]
 pub enum AnyOfObjectAnyOf {
     #[serde(rename = "FOO")]
-    FOO,
+    Foo,
```
Comment (jacob-pro) [inline samples/server/petstore/rust-server/output/openapi-v3/src/models.rs:632]: @wing328 

>  sounds like that's a breaking change (not backward-compatible) right?

Yes that is correct, as you can see enum variants have changed to camel case in the rust server to comply with the Rust guidelines https://rust-lang.github.io/api-guidelines/naming.html

For example:

```
enum ExampleEnum {
    VARIANT_ONE
}
```

If we don't make this change then warnings are emitted:

```
warning: variant `VARIANT_ONE` should have an upper camel case name
  --> reqwest\petstore\src\lib.rs:13:5
   |
13 |     VARIANT_ONE
   |     ^^^^^^^^^^^ help: convert the identifier to upper camel case: `VariantOne`
   |
   = note: `#[warn(non_camel_case_types)]` on by default
```

Comment (jacob-pro) [reply samples/server/petstore/rust-server/output/openapi-v3/src/models.rs:632]: Edit: I can see that `rust-server` is supressing these warnings in lib.rs

**Comment by jacob-pro** [comment]
> FYI. To run the tests locally, you will need to run a local petstore server: https://github.com/OpenAPITools/openapi-generator/wiki/Integration-Tests#how-to-add-integration-tests-for-new-petstore-samples

I am able to successfully build `samples/client/petstore/rust` locally.

Trying to run `rust-server` is proving more difficult.

First there is a bug that prevents me compiling on Windows. Cargo doesn't seem to support the `git://`  notation, so the templates should really use the `https://` instead which does work:

```
$ cargo build
    Updating git repository `git://github.com/Metaswitch/serde-xml-rs.git`
warning: spurious network error (2 tries remaining): failed to connect to github.com: ; class=Os (2)


jacob@JACOB-PC MINGW64 ~/IdeaProjects/test-git
$ cargo build
    Updating git repository `https://github.com/Metaswitch/serde-xml-rs.git`
```

Then I am having other issues due to this being set: `"-D", "warnings", # all warnings should be denied`

```
   Compiling openapi-v3 v1.0.7 (C:\Users\jacob\IdeaProjects\openapi-generator\samples\server\petstore\rust-server\output\openapi-v3)
error: field is never read: `nullable`
    --> output\openapi-v3\src\models.rs:1647:13
     |
1647 |             pub nullable: Vec<String>,
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: `-D dead-code` implied by `-D warnings`
```

I don't however believe these errors have anything to do with my changes

**Comment by wing328** [comment]
rust client, server tests passed via https://app.travis-ci.com/github/OpenAPITools/openapi-generator/builds/254830850

File: samples/server/petstore/rust-server/output/openapi-v3/src/models.rs (hunk 1 above)
Comment (wing328) [reply samples/server/petstore/rust-server/output/openapi-v3/src/models.rs:632]: if the client prefers the old way of naming the enum, we will add an option for fallback later. Let's see if there's any demand for this from rust users after the v6.1.0 release.

Comment (jacob-pro) [reply samples/server/petstore/rust-server/output/openapi-v3/src/models.rs:632]: thanks, if this does come up - just tag me and I can write the code for you

**Pull request merged**